lib/compress/zstdmt_compress.c

```diff
@@ -1089,29 +1110,20 @@ static size_t ZSTDMT_createCompressionJob(ZSTDMT_CCtx* mtctx, size_t srcSize, ZS
         mtctx->jobs[jobID].frameChecksumNeeded = endFrame && (mtctx->nextJobID>0) && mtctx->params.fParams.checksumFlag;
         mtctx->jobs[jobID].dstFlushed = 0;
 
-        if (mtctx->params.fParams.checksumFlag)
-            XXH64_update(&mtctx->xxhState, (const char*)mtctx->inBuff.buffer.start + mtctx->inBuff.prefixSize, srcSize);
+        if (mtctx->params.fParams.checksumFlag && srcSize > 0)
+            XXH64_update(&mtctx->xxhState, src, srcSize);
 
-        /* get a new buffer for next input */
+        /* Update the round buffer pos and clear the input buffer to be reset */
+        mtctx->inBuffPos += srcSize;
+        mtctx->inBuff.buffer = g_nullBuffer;
+        mtctx->inBuff.filled = 0;
+        /* Set the prefix */
         if (!endFrame) {
-            size_t const newPrefixSize = MIN(mtctx->inBuff.filled, mtctx->targetPrefixSize);
-            mtctx->inBuff.buffer = ZSTDMT_getBuffer(mtctx->bufPool);
-            if (mtctx->inBuff.buffer.start == NULL) {    /* not enough memory to allocate a new input buffer */
-                mtctx->jobs[jobID].srcSize = mtctx->jobs[jobID].consumed = 0;
-                mtctx->nextJobID++;
-                ZSTDMT_waitForAllJobsCompleted(mtctx);
-                ZSTDMT_releaseAllJobResources(mtctx);
-                return ERROR(memory_allocation);
-            }
-            mtctx->inBuff.filled -= (mtctx->inBuff.prefixSize + srcSize) - newPrefixSize;
-            memmove(mtctx->inBuff.buffer.start,   /* copy end of current job into next job, as "prefix" */
-                (const char*)mtctx->jobs[jobID].prefixStart + mtctx->inBuff.prefixSize + srcSize - newPrefixSize,
-                mtctx->inBuff.filled);
-            mtctx->inBuff.prefixSize = newPrefixSize;
+            size_t const newPrefixSize = MIN(srcSize, mtctx->targetPrefixSize);
```

Comment:
If an intermediate job is small (due to a `flush()` for example),
then the prefix of next job is going to be limited.

It seems there is no more any good reason for that limitation.
Since all input lies within a single big buffer, a prefix could encompass multiple previous jobs.
The comparison should rather be with the amount of data available within input buffer, hence : 

`size_t const newPrefixSize = MIN(inBuff.filled, mtctx->targetPrefixSize);`

note: I'm not sure at this stage if `inBuff` is really about the large single buffer.

Reply:
I'm going to leave this until later, I created a task with the list of minor improvement follow-ups.